### ISSUE TITLE
When specifying an ENI, the module breaks

### ISSUE THREAD

**Issue by bgray-onica** [terraform-aws-modules/terraform-aws-ec2-instance#188]
This is something I don't honestly know how to fix in a generic way, but this module breaks when specifying an ENI. The only way to fix is to remove line 27 in `main.tf`. Maybe have two versions where one is instantiated of the ENI is defined and one that is instantiated if the ENI is not defined.

```
Error: ConflictsWith

  on .terraform/modules/myserv er/main.tf line 5, in resource "aws_instance" "this":
   5: resource "aws_instance" "this" {

"network_interface": conflicts with associate_public_ip_address
```

**Comment by antonbabenko** [comment]
You are right, there is [a note about this specific limitation](https://github.com/terraform-aws-modules/terraform-aws-ec2-instance#notes) which was there since the initial version of the module, but can be improved in the recent versions of Terraform.

Pull-request is welcome.

**Comment by jeremyciak** [comment]
I would love to help with a PR. I'm a bit confused by the situation and the proposed fix. Line 27 is:

```terraform
ebs_optimized = var.ebs_optimized
```

Why would that line have anything to do with ENI? Also, I have a separate version of this module that I maintain myself where I have split the ENI's off entirely so that you can make ENI changes without needing to recreate the EC2 instance. Is there any interest in adopting that for this module?

**Comment by bgray-onica** [comment]
Oops I did mean line 22. Sorry about that.

That may be a good approach but I disagree with having 6732461982346972134 separate modules just because of some specific needs.

Here's an example we could work with: https://github.com/terraform-aws-modules/terraform-aws-alb. So previously, this module in 0.11 and older versions made for 0.12 had a tf for `with access logs` and one `no access logs` because they were not able to be commingled previously. Now they've moved to this solution:

```
  dynamic "access_logs" {
    for_each = length(keys(var.access_logs)) == 0 ? [] : [var.access_logs]

    content {
      enabled = lookup(access_logs.value, "enabled", lookup(access_logs.value, "bucket", null) != null)
      bucket  = lookup(access_logs.value, "bucket", null)
      prefix  = lookup(access_logs.value, "prefix", null)
    }
  }
```

I'll play around with it and see if I can't figure something out, but this seems like it would be a pretty great option instead. That way it can be included or not dynamically.

**Comment by jeremyciak** [comment]
Is my PR #189 insufficient or is this all that is needed?

**Comment by bgray-onica** [comment]
I tried just setting it to null and it fails still. Same error. The option needs to not exist at all or else it errors out.

**Comment by jeremyciak** [comment]
@bgray-onica, are you setting the primary network interface with the `network_interface` variable? What is your variable composition that you are testing with?

**Comment by jeremyciak** [comment]
Although I'm seeing now that the default value for `associate_public_ip_address` is already `null` so my logic doesn't really accomplish anything. Back to the drawing board... but if I could get insight into your composition for the configuration use-case you're working towards that would be helpful.

**Comment by jeremyciak** [comment]
I think we are actually still limited by the `aws` provider here with the `aws_instance` resource:

https://github.com/hashicorp/terraform-provider-aws/blob/master/aws/resource_aws_instance.go

The logic around handling the `associate_public_ip_address` attribute is simply being consumed by this module and I don't know that we can work our way around it. I think we may need to submit this issue at the `aws` provider instead of here for this module, but I welcome other input.
